### PR TITLE
Implement Automatic Mixed Precision with GradScaler to Address NaN Loss Issues

### PR DESCRIPTION
**Pull request by aiihn**
### Description

This pull request addresses the issue of NaN losses occurring during mixed-precision training with `--fp16` enabled (#12).

### Key Changes

- Integrated `torch.cuda.amp.GradScaler` to dynamically adjust loss scaling.
- Replaced the manual loss scaling approach. Note: `GradScaler` will override the `loss_scale` set manually by `--ls`.

### Usage

Use `--fp16=True` along with `--enable_gradscaler=True`. For example, below is the mixed-training command modified from [run_ecm_1hour.sh](https://github.com/locuslab/ect/blob/cb3536063a51b9fc1d8e64dccb1cc2f448ab0fa6/run_ecm_1hour.sh).
```
torchrun --nnodes=1 --nproc_per_node=1 --rdzv_backend=c10d --rdzv_endpoint=localhost:9901 ct_train.py  \
    --outdir=ct-runs --data=datasets/cifar10-32x32.zip  \
    --cond=0 --arch=ddpmpp --metrics=fid50k_full        \
    --transfer=https://nvlabs-fi-cdn.nvidia.com/edm/pretrained/edm-cifar10-32x32-uncond-vp.pkl    \
    --duration=25.6 --tick=12.8 --double=250 --batch=128 --lr=0.0001 --optim=RAdam --dropout=0.2 --augment=0.0 \
    -q 256 --double 10000 --ema_beta 0.9993 --eval_every 80 --dump 80     \
    --desc bs128.200k \
    --fp16=True --enable_gradscaler=True
```
The FID records obtained using the above command are shown in the following images:
image
image

### CODE REVIEW

File: ct_train.py [hunk 1]
```diff
@@ -78,6 +78,7 @@ def convert(self, value, param, ctx):
 @click.option('--fp16',          help='Enable mixed-precision training', metavar='BOOL',            type=bool, default=False, show_default=True)
 @click.option('--tf32',          help='Enable tf32 for A100/H100 training speed', metavar='BOOL',   type=bool, default=False, show_default=True)
 @click.option('--ls',            help='Loss scaling', metavar='FLOAT',                              type=click.FloatRange(min=0, min_open=True), default=1, show_default=True)
+@click.option('--enable_gradscaler', help='Enable torch.cuda.amp.GradScaler, NOTE overwritting loss_scale set by --ls', metavar='BOOL', type=bool, default=False, show_default=True)
```
Comment (Gsunshine) [inline ct_train.py:81]: Hi Zixiang @aiihn ,

Thanks for your neat PR!

Would it be better to use a short abbreviation like `amp` as the option name? AMP already stands for Automatic Mixed Precision.

File: training/ct_training_loop.py [hunk 2]
```diff
@@ -197,6 +206,13 @@ def training_loop(
         data = torch.load(resume_state_dump, map_location=torch.device('cpu'))
         misc.copy_params_and_buffers(src_module=data['net'], dst_module=net, require_all=True)
         optimizer.load_state_dict(data['optimizer_state'])
+        if enable_gradscaler:
+            if 'gradscaler_state' in data:
+                dist.print0(f'Loading GradScaler state from "{resume_state_dump}"...')
+                # Although not loading the state_dict of the GradScaler works well, loading it can improve reproducibility.
```
Comment (Gsunshine) [inline training/ct_training_loop.py:212]: Gotcha. Thanks for the comments!

File: training/ct_training_loop.py [hunk 3]
```diff
@@ -253,16 +269,24 @@ def update_scheduler(loss_fn):
...
+                else:
+                    # loss.sum().mul(loss_scaling / batch_gpu_total).backward()
+                    loss.mul(loss_scaling).mean().backward()
+        
+        if enable_gradscaler:
+            # TODO Is torch.nan_to_num(param.grad, nan=0, posinf=1e5, neginf=-1e5, out=param.grad) needed when using GradScaler?
+            scaler.step(optimizer)
+            scaler.update()
+        else:
+            # Update weights.
```
Comment (Gsunshine) [inline training/ct_training_loop.py:283]: TODO is also unclear to me either. It seems still useful and compatible per Claude.

It's fine to remove my commented code for lr rampup.

**Review by Gsunshine — APPROVED** [review]
Merge AMP via Gradscalar into ECT.

**Comment by Gsunshine** [comment]
Hi @aiihn ,

Thank you again for your PR! I had another AMP implementation that could also be helpful for ECT. I’ll check it out later and test `torch.autocast`, bu feel free to take a look if you’re working with mixed precision!

Links for reference:
https://github.com/locuslab/torchdeq/blob/main/deq-zoo/deq-flow/main.py
https://github.com/locuslab/torchdeq/blob/main/deq-zoo/deq-flow/core/deq_flow.py

Cheers,
Zhengyang
